src/ripple/app/misc/impl/AMM.cpp

```diff
@@ -0,0 +1,248 @@
+        if (optIssue1 && optIssue2)
+            return {{*optIssue1, *optIssue2}};
+        auto const issue1 = ammSle[sfAsset];
+        auto const issue2 = ammSle[sfAsset2];
+        if (optIssue1)
+        {
+            if (*optIssue1 == issue1)
+                return std::make_optional(std::make_pair(issue1, issue2));
+            else if (*optIssue1 == issue2)
+                return std::make_optional(std::make_pair(issue2, issue1));
+            JLOG(j.debug()) << "ammHolds: Invalid optIssue1 " << *optIssue1;
+            return std::nullopt;
+        }
+        else if (optIssue2)
+        {
+            if (*optIssue2 == issue2)
+                return std::make_optional(std::make_pair(issue2, issue1));
+            else if (*optIssue2 == issue1)
+                return std::make_optional(std::make_pair(issue1, issue2));
+            JLOG(j.debug()) << "ammHolds: Invalid optIssue2 " << *optIssue2;
+            return std::nullopt;
+        }
+        return std::make_optional(std::make_pair(issue1, issue2));
+    }();
+    if (!issues)
+        return Unexpected(tecAMM_INVALID_TOKENS);
+    auto const [asset1, asset2] = ammPoolHolds(
+        view,
+        ammSle.getAccountID(sfAMMAccount),
+        issues->first,
+        issues->second,
+        freezeHandling,
+        j);
+    return std::make_tuple(asset1, asset2, ammSle[sfLPTokenBalance]);
+}
+
+STAmount
+ammLPHolds(
+    ReadView const& view,
+    Currency const& cur1,
+    Currency const& cur2,
+    AccountID const& ammAccount,
+    AccountID const& lpAccount,
+    beast::Journal const j)
+{
+    auto const lptIss = ammLPTIssue(cur1, cur2, ammAccount);
+    return accountHolds(
+        view,
+        lpAccount,
+        lptIss.currency,
+        lptIss.account,
+        FreezeHandling::fhZERO_IF_FROZEN,
+        j);
+}
+
+STAmount
+ammLPHolds(
+    ReadView const& view,
+    SLE const& ammSle,
+    AccountID const& lpAccount,
+    beast::Journal const j)
+{
+    return ammLPHolds(
+        view,
+        ammSle[sfAsset].currency,
+        ammSle[sfAsset2].currency,
+        ammSle[sfAMMAccount],
+        lpAccount,
+        j);
+}
+
+bool
+isFrozen(ReadView const& view, STAmount const& a)
+{
+    return !a.native() && isGlobalFrozen(view, a.getIssuer());
+}
+
+TER
+requireAuth(ReadView const& view, Issue const& issue, AccountID const& account)
+{
+    if (isXRP(issue) || issue.account == account)
+        return tesSUCCESS;
+    if (auto const issuerAccount = view.read(keylet::account(issue.account));
+        issuerAccount && (*issuerAccount)[sfFlags] & lsfRequireAuth)
+    {
+        if (auto const trustLine =
+                view.read(keylet::line(account, issue.account, issue.currency));
+            trustLine)
+            return !((*trustLine)[sfFlags] &
+                     ((account > issue.account) ? lsfLowAuth : lsfHighAuth))
+                ? TER{tecNO_AUTH}
+                : tesSUCCESS;
```

Comment:
Looking at this again, I think it would be a touch easier to read and reason about if you remove the `!` and reverse the results. So:
```
            return (*trustLine)[sfFlags] &
                     ((account > issue.account) ? lsfLowAuth : lsfHighAuth)
                ? tesSUCCESS
                : TER{tecNO_AUTH};
```